### DISCUSSION TITLE
Import albums from old library into new library?

### DISCUSSION THREAD

**Discussion by alexklibisz** [RhetTbull/osxphotos#1620] [General]
Hi, thanks for building this tool!

I'm trying to update my Photos library so that the originals are stored outside the library. In my case, I'd like for the originals to live on a NAS, as I'm running out of space on my Macbook.

I know that I can create a new library and re-import photos with this setting disabled:

<img width="437" alt="image" src="https://github.com/user-attachments/assets/25eee4bf-c1d6-404c-b8a3-099d9aab436a">

I've figured out how to use the `osxphotos export` command to export all originals to a reasonable folder structure, and I can import them to a new Photos library with that setting unchecked. That works well.

However, I would also like to preserve albums, and I can't figure out how to do that after a couple hours of fiddling with the commands.

In short, I want this:

* Before: my library contains all of the originals with some of them sorted into albums.
* After: my library does not contain the originals (they are stored elsewhere and imported to the library), but the photos are still sorted into the same albums.

Or maybe there's another way to approach this? Maybe there's a way to edit the existing library to set the location of the file to my new external location? I'm open to ideas!

**Comment by oPromessa** [comment]
### Preamble
1. All my ~100K pics are "Referenced" located on the local disk (Yes, you need space locally!).
   * I use osxphotos to export to Synology NAS and use Synology Photos to access them from anywhere (mobile, browsers, etc). 
2. Don't have experience with "Referenced" Files on NAS. **But do consider the following:**
   * See [Photos on NAS as referenced files - any issues and library size](https://discussions.apple.com/thread/251316120?sortBy=rank)
   * RhetTbull also strongly advises against the use of "Referenced" files:
> As an aside, I strongly advise to not use a referenced file. If you ever change Macs or change the drive the files are on (even if path is the same), then Photos will fail to find the originals. This is because Photos uses security-scoped bookmarks that are sand-boxed to the original volume ID and path to find the originals. I have a project to fix this ([RepairPhotosBookmarks](https://github.com/RhetTbull/RepairPhotosBookmarks)) that I intend to eventually incorporate in OSXPhotos but rewriting the database takes significant care and I don't have time at the moment to work on this. The repair code is thus "use as is / best effort". It may or may not work with your version of macOS. Use referenced libraries with caution!


### If you're still going ahead....
* Check out #1267 
* RhetTbull made recently an excellent feature [`osxphotos import --exportdb`](https://rhettbull.github.io/osxphotos/cli.html#osxphotos-import) that will **recreate albums** and **recover most of the metadata available in the previous library** by making use of the osxphotos export database (from a previous export run). 

#### 1. export
* **OPEN YOUR CURRENT LIBRARY ON PHOTOS**. 
* Do see the applicability of each option for your case. I use --not-hidden for instance.
* **Update**: Use `--export-by-date` to create output folders to organize photos by date created (e.g. DEST/2019/12/20/photoname.jpg).
```py
osxphotos export /Volumes/NASVolume/export --export-by-date --cleanup --update --exiftool --touch-file --exiftool_merge_persons --exiftool_merge_keywords --sidecar xmp --export_aae --not-hidden --keyword_template "{keyword}" --ramdb --retry 2 --skip_edited --verbose
```
#### 2. import
* **MAKE SURE YOU CLOSE THE CURRENT LIBRARY IN PHOTOS. AND OPEN THE ONE YOU WANT TO IMPORT INTO**. 
* Import will import into the last opened Photos Library.
* Review and select the appropriate options. I use --resume for instance.
* Please consider using `--stop-on-error 5` as my experience has been that every 2K pics imported AppleScript get's blocked in an error-loop and prevents `osxphotos import` from continuing... so it might as well stop it. Restart Photos and re-run `osxphotos import`.
```py
osxphotos import /Volumes/NASVolume/export --exportdb /Volumes/NASVolume/export/.osxphotos_export.db --walk --stop-on-error 5 --sidecar --verbose --timestamp  --dup-check --skip-dups --resume

**Comment by RhetTbull** [comment]
Hi @alexklibisz and thanks @oPromessa for the detailed response. 

OSXPhotos should be able to help. You can use `osxphotos sync` to export the metadata from the old library and apply it to the library. OSXPhotos will match the photos via fingerprint (a hash of each photo) and apply the new metadata. 

To see the help on sync: `osxphotos help sync` or `osxphotos docs`.

On the old library:

`osxphotos sync --export old_library_metadata.db --library /path/to/old/library`

Then open the new library in Photos and run:

`osxphotos sync --import old_library_metadata.db --merge albums` 

You can also merge (or replace with `--set` instead of `--merge`) keywords, descriptions (captions), titles, and favorite status. See the `osxphotos help sync merge` for more info.

As @oPromessa mentioned, you could theoretically also use the `osxphotos import` command to recreate the library. However, I've noticed a new bug (#1621) when I was testing my initial response to this question that results in Photos copying images that have been edited to the Photos library instead of maintaining the reference so I would avoid using this solution for now.  For the case where you want to preserve edits, OSXPhotos can faithfully recreate the library but NOT if it is a referenced library due to a limitation in Photos -- imported edited images to Photos causes them to be copied even if the "don't copy" flag is set.

A few comments on references:

1. I strongly advise against using referenced libraries. If you ever switch Macs or move the photos to a different volume (even one with the same name), Photos will not be able to find the referenced images due to App sandboxing and the way Photos uses security scoped (sandboxed) bookmarks to store the path to the destination file.
2. I strongly advise against storing any data Photos will use on a NAS. You must certainly never put the actual library on the NAS (even Apple advises against this) due to the way Photos accesses the internal SQLite data store (`Photos.db` in the library) which is not compatible with NAS storage.
3. If you want to proceed, I recommend using `osxphotos sync` to sync the metadata.

I have a partial fix ([RepairPhotosBookmarks](https://github.com/RhetTbull/RepairPhotosBookmarks)) as @oPromessa pointed out that will address #1 and I plan to eventually add this to OSXPhotos but it's not yet 100% reliable.

> Or maybe there's another way to approach this? Maybe there's a way to edit the existing library to set the location of the file to my new external location? I'm open to ideas!

Unfortunately this isn't possible due to the way Photos uses security scoped bookmarks as mentioned in #1. It's theoretically  possible once I get the RepairPhotosBookmarks code integrated into Photos but I would not advise doing so. Even once fully implemented I expect this feature will be  more of an emergency repair type option than something I recommend for routine use. It requires a pretty ugly hack to make it work.

**Reply by oPromessa** [reply]
> OSXPhotos should be able to help. You can use `osxphotos sync` to export the metadata from the old library and apply it to the library. OSXPhotos will match the photos via fingerprint (a hash of each photo) and apply the new metadata.

Quick question @RhetTbull. Sync will sync the metadata in two Photos Libraries. But, in this case, one stll needs to:
* have the pics moved "out of" the old Photos Library into a new location (local disk, NAS, etc)
* and have the pics imported into the new Photos Library
prior to running **sync**, or am I seeing it wrong?

**Reply by RhetTbull** [reply]
@oPromessa yes. sync would only work if you'd already imported the photos into a new library.

**Comment by alexklibisz** [comment]
Thank you both for the thorough responses. I'll try them out and I'll re-consider whether the reference storage is a good option. I'll re-open this if I have any more questions.
